architecture/2025-02-user-defined-resource-type-feature-spec.md

```diff
@@ -0,0 +1,1212 @@
+                  description: "Required: The container image"
+                ...
+```
+
+**User Experience 1 – Using Terraform configuration**
+
+```diff
+terraform {
+  required_providers {
+    radius = {
+      source  = "terraform.radapp.io"
+      version = "~> 0.46"
+    }
+  }
+}
+
+variable "context" {
+  description = "This variable contains Radius recipe context."
+  type = any
+}
+
++ if var.context.resource.properties.ingress {
++  resource "applications_core_gateways" "gateway" {
+    name                = "gateway"
+    environment         = var.context.environment.id
+    application         = var.context.application
+    routes = {
+      path              = "/"
+      destination       = "http://${local.serviceContainer.name}:${context.resource.properties.container.ports.http.containerPort}'
+    }     
+
+  }
+}
+
+resource "applications_core_containers" "serviceContainer" {
+  name                = "${var.context.resource.name}-container"
+  environment         = var.context.environment.id
+  application         = var.context.application.id
+  container_image     = context.resource.properties.container.image
+  ...
+}
+```
+
+**User Experience 2 – Using Bicep template**
+
+The recipe will have a conditional on the ingress property to determine whether to deploy a gateway resource.
+
+**`webservice-recipe.bicep`**:
+
+```diff
+extension radiusCore
+
+@description('Information about what resource is calling this Recipe. Generated by Radius.')
+param context object
+
++resource gateway 'Applications.Core/gateways@2023-10-01-preview' = if (context.resource.properties.ingress) {
+  name: 'gateway'
+  properties: {
+    application: context.application.id
+    routes: [
+      {
+        path: '/'
+        destination: 'http://${serviceContainer.name}:${context.resource.properties.container.ports.http.containerPort}'
+      }
+    ]
+  }
+}
+
+resource serviceContainer 'Applications.Core/containers@2023-10-01-preview' = {
+  name: '${context.resource.name}-container'
+  properties: {
+    application: context.application.id
+    container: context.resource.properties.container
+    ...
+  }
+}
+```
+
+### User Story 7 – Registering recipes
+
+As a platform engineer, I need to register a recipe which implements my new resource type. 
+
+**Summary**
+
+Registering recipes works as expected just like existing portable resource types. However, since user-defined resource types introduces new capabilities, there are adjustments made to the recipe capabilities of Radius.
+
+**User Experience**
+
+```bash
+# Register the PostgreSQL Terraform recipe in the myEnv environment
+rad recipe register \
+  --environment myEnv \
+  --resource-type Radius.Resources/postgreSQL \
+  --recipe-kind terraform \
+  --recipe-path git://github.com/<username>/repo.git//recipes/postgreSQL
+Registering recipe for Radius.Resources/postgreSQL
+The recipe for Radius.Resources/postgreSQL is registered in the myEnv environment
+```
```

Comment:
Is there any sort of validation in this flow? Like the validation of the recipe in the given path or at least if there is a recipe in the given path?